File: 76.py
from collections import Counter

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(t) == 0:
            return 0
        res = [-1, -1]
        min_len = len(s) + 1
        l = 0

        count_t = Counter(t)
        to_match = len(count_t)
        matches = 0
        count_s = Counter()

        for index, letter in enumerate(s):
            count_s[letter] += 1
            if(count_s[letter] == count_t[letter]):
                matches += 1

            while(matches == to_match):
                size = index - l + 1
                if(size < min_len):
                    min_len = size
                    res = [l, index + 1]
                if(count_s[s[l]] == count_t[s[l]]):
                    matches -= 1
                count_s[s[l]] -= 1
                l += 1

        return s[res[0]: res[1]] if min_len != len(s) + 1 else ""

File: 76.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> count_s, count_t;
        int left = 0, smallest = s.size() + 1, start, matches = 0;
        int count;

        for(int i = 0; i < t.size(); i++) {
            count_t[t[i]]++;
        }

        count = count_t.size();
        for(int i = 0; i < s.size(); i++) {
            count_s[s[i]]++;
            if(count_s[s[i]] == count_t[s[i]]) {
                matches += 1;
            }
            while(matches == count) {
                if(smallest > i - left + 1) {
                    smallest = i - left + 1;
                    start = left;
                }
                if(count_s[s[left]] == count_t[s[left]]) {
                    matches--;
                }
                count_s[s[left]]--;
                if(count_s[s[left]] == 0) {
                    count_s.erase(s[left]);
                }
                left++;
            }
        }
        if(smallest == s.size() + 1) {
            return "";
        }
        return s.substr(start, smallest);
    }
};